### PR TITLE
Bug/subform validation

### PR DESCRIPTION
**Pull request by Sven65**
## Explain the changes you’ve made

In accordance with #CU-1exphyp, I've made changes to not let users save subforms without entering values.

## Explain why these changes are made

As users shouldn't save a field without entering a value, these changes are made.

## Explain your solution

In order to facilitate checking the subform fields in the easiest way, I've added a way to know when a user has added a new answer to a `RepeaterField` and added a handler of that event which sets the `RepeaterField` to dirty and performs a validation on it.

As the user shouldn't be able to save the subform with empty values, I've changed the `backToMain` case in StepFooter to also check for errors before allowing the user to move on.

In this pull request, I've also made the `RepeaterField` answers show their validation errors.

## How to test the changes?

Concrete example:
1. Checkout this branch
2. Start a case for "Ensamsökande"
3. Try adding an empty field in "medicin och sjukvård".
4. See that it doesn't let you

## Was this feature tested in the following environments?
- [] Storybook on a iOS device/simulator.
- [x] Building the Application on a iOS device/simulator.
- [] Building the Application on a Android device/simulator.

## Anything else?

In order for this to work, we're going to need some changes to be made using the form builder.
More specifically, we need to set the inputs rows on repeater fields to have required answers.

### CODE REVIEW

**Comment by helsingborgsstad** [comment]
Task linked: [CU-1exphyp Valideringsfel i subform - ett fält kan sparas utan värde](https://app.clickup.com/t/1exphyp)

**Comment by KenthLjungHBG** [comment]
It doesn't seem to work, or does this require an updated form as well?

**Comment by Sven65** [comment]
> It doesn't seem to work, or does this require an updated form as well?

You need to update the form using the formbuilder. Set the subform inputs to required.

File: source/components/molecules/RepeaterField/RepeaterFieldListItem.tsx [hunk 1]
```diff
@@ -191,6 +199,8 @@ const RepeaterFieldListItem: React.FC<Props> = ({
             <InputComponent 
               input={input} 
               onChange={changeFromInput(input)} 
+              error={error && error[input.id] ? {isValid: error[input.id].isValid, message: error[input.id].validationMessage} : undefined}
+              showErrorMessage={error && error[input.id] && !error[input.id].isValid}
```
Comment (KenthLjungHBG) [inline source/components/molecules/RepeaterField/RepeaterFieldListItem.tsx:203]: This is a bit wordy for an inline-expression. Could we extract this to two variables, one for the error object and a flag for if we want to show an error respectively?

File: source/components/molecules/RepeaterField/RepeaterField.tsx [hunk 2]
```diff
@@ -27,6 +27,7 @@ interface Props {
   value: string | Record<string, string | number>[];
   onChange: (answers: Record<string, any> | string | number, fieldId?: string) => void;
   onBlur?: (answers: Record<string, any> | string | number, fieldId?: string) => void;
+  onAddAnswer?: (answers: Record<string, any> | string | number, fieldId?: string) => void;
```
Comment (karlknowit) [inline source/components/molecules/RepeaterField/RepeaterField.tsx:30]: A little DRY perhaps?
answers could be its own type.

```js
type Answers = Record<string, any> | string | number;
```

and use like:
```js
onAddAnswer?: (answers: Answers, fieldId?: string) => void;
```

**Review by KenthLjungHBG — CHANGES_REQUESTED** [review]
Got it working as expected by updating the form, just had a minor refactor note 🙂. Also I notice the files aren't conforming to the new styling setup - do you have eslint/prettier extensions setup? Re-styling the files might introduce too much noise so I'll maybe look into a different solution for that.

File: source/components/molecules/RepeaterField/RepeaterFieldListItem.tsx [hunk 3]
```diff
@@ -191,6 +199,8 @@ const RepeaterFieldListItem: React.FC<Props> = ({
             <InputComponent 
               input={input} 
               onChange={changeFromInput(input)} 
+              error={error && error[input.id] ? {isValid: error[input.id].isValid, message: error[input.id].validationMessage} : undefined}
```
Comment (karlknowit) [inline source/components/molecules/RepeaterField/RepeaterFieldListItem.tsx:202]: You could extract this into two constants instead like:

```js
const error = error?.[input.id] ? {isValid: error[input.id].isValid, message: error[input.id].validationMessage} : undefined
const showErrorMessage = !error?.[input.id]?.isValid
```

File: source/components/organisms/Step/StepFooter/StepFooter.tsx [hunk 4]
```diff
@@ -83,7 +83,15 @@ const StepFooter: React.FC<Props> = ({
         return onSubmit || null;
       }
       case 'backToMain': {
-        return formNavigation.goToMainForm;
+        return () => {
+          const errorCallback = () => {};
+  
+          const onValidCallback = async () => {
```
Comment (karlknowit) [inline source/components/organisms/Step/StepFooter/StepFooter.tsx:89]: Why does this function has to be async when we aren't awaiting anything?

File: source/containers/FormField/FormField.js [hunk 5]
```diff
@@ -212,6 +219,8 @@ const FormField = ({
   if (input && input.focusEvent) inputCompProps[input.focusEvent] = onInputFocus;
   if (input && input.helpInComponent) inputCompProps[input.helpProp || 'help'] = help;
 
+  if (input && inputType === 'repeaterField' && input.onAddAnswerEvent) inputCompProps[input.onAddAnswerEvent] = onInputAddAnswer;
```
Comment (karlknowit) [inline source/containers/FormField/FormField.js:222]: a little less code perhaps:

```js
if(inputType === 'repeaterField' && !!input?.onAddAnswerEvent) inputCompProps[input.onAddAnswerEvent] = onInputAddAnswer;
```

File: source/components/organisms/Step/StepFooter/StepFooter.tsx [hunk 6]
```diff
@@ -136,17 +144,17 @@ const StepFooter: React.FC<Props> = ({
...
-        onClick={actionMap(action)}
-        colorSchema={action.color}
-        disabled={isLoading || (action.hasCondition && checkCondition(action.conditionalOn))}
-        z={0}
-      >
-        <Text>{action.label}</Text>
-      </Button>
-    </Flex>
-  ));
+        <Flex key={`${index}-${action.label}`}>
```
Comment (karlknowit) [inline source/components/organisms/Step/StepFooter/StepFooter.tsx:147]: It isn't best practices to use **index** in the key property since that could possibly affect the React performance.

Comment (Sven65) [reply source/components/organisms/Step/StepFooter/StepFooter.tsx:147]: I'm not sure if this would break things, and I'm not sure if I should change it, as I don't see any other simple way of making the key unique.

File: source/components/organisms/Step/StepFooter/StepFooter.tsx (hunk 4 above)
Comment (Sven65) [reply source/components/organisms/Step/StepFooter/StepFooter.tsx]: Great question, fixed!

File: source/components/organisms/Step/StepFooter/StepFooter.tsx (hunk 6 above)
Comment (KenthLjungHBG) [reply source/components/organisms/Step/StepFooter/StepFooter.tsx:147]: Agree, looking at the code a possible value could be `${action.type}-${action.label}` as it should be sufficiently unique.

**Comment by Sven65** [comment]
> Got it working as expected by updating the form, just had a minor refactor note 🙂. Also I notice the files aren't conforming to the new styling setup - do you have eslint/prettier extensions setup? Re-styling the files might introduce too much noise so I'll maybe look into a different solution for that.

I've got the eslint extension for vscode setup to automatically format code on save. I haven't setup prettier for vscode as it conflicted with eslint last time I tried it due to how they were configured in the repo.

File: source/containers/FormField/FormField.js (hunk 5 above)
Comment (KenthLjungHBG) [reply source/containers/FormField/FormField.js]: I do not like double nots in javascript. It's a very code-golfy javascript-y way of forcing a boolean, although I can see the appeal here.

Looking into this code it's a bit it's very messy. If I'm not mistaken it exits out early on line 178 if `input === undefined` meaning that for the rest of the code there should be no reason to check `if(input && ..` at all (unless it's somehow `null` instead of `undefined`.

For now I would just replace this line with `if(input.onAddAnswerEvent)` since we probably don't need to check for a repeater field anyway; if `onAddAnswerEvent` is specified then we want it regardless of input type right?

File: source/containers/FormField/FormField.js [hunk 7]
```diff
@@ -128,6 +128,7 @@ const inputTypes = {
     focusEvent: 'onFocus',
     blurEvent: 'onBlur',
     changeEvent: 'onChange',
+    onAddAnswerEvent: 'onAddAnswer',
```
Comment (KenthLjungHBG) [inline source/containers/FormField/FormField.js:131]: Noticed now but this is using a different naming scheme than the rest. In that case it should be renamed to `addAnswerEvent`.

**Review by KenthLjungHBG — CHANGES_REQUESTED** [review]
Found a minor inconsistency. I also had some notes on the new FormField change which I outlined in a comment on the thread there instead 🙂.

**Review by dannil76 — APPROVED** [review]

**Review by dannil76 — APPROVED** [review]
